### PR TITLE
Allow users to use NULL for pThingName, pThingNameLength when invoke …

### PR DESCRIPTION
**Pull request by DanielYEHsieh**
1. Allow users to use NULL for pThingName, pThingNameLength when invoke Shadow_MatchTopic If they are not interested in pThingName, pThingNameLength. (Align Jobs Match topic style)

2. Made related changes to reduce complexity score after adding change#1 .

3. Fixed CI failure due to path changes libraries/aws/shadow --> libraries/aws/device-shadow-for-aws-iot-embedded-sdk

### CODE REVIEW

File: source/shadow.c [hunk 1]
```diff
@@ -466,37 +463,34 @@ ShadowStatus_t Shadow_MatchTopic( const char * pTopic,
...
-                LogDebug( ( "Not related to Shadow, thing name is not in pTopic %s, failed to parse thing name", pTopic ) );
-            }
-        }
-        else
+        /* If no more topic string is left to parse, fail. */
+        if( consumedTopicLength >= topicLength )
         {
-            LogDebug( ( "Not related to Shadow, failed to parse shadow topic prefix in pTopic %s", pTopic ) );
+            shadowStatus = SHADOW_THINGNAME_PARSE_FAILED;
+            LogDebug( ( "Not related to Shadow, thing name is not in pTopic %s, failed to parse thing name", pTopic ) );
```
Comment (sarenameas) [inline source/shadow.c:476]: These should be LogError()?

Comment (sarenameas) [reply source/shadow.c:476]: per offline conversation, it is not an error to fail to parse.

Comment (DanielYEHsieh) [reply source/shadow.c:476]: We have the status code to notify users that this is not a shadow topic.
PARSE fail means this is not a shadow topic, it doesn't means there is a error that needs to be handle

**Review by sarenameas — APPROVED** [review]

File: source/shadow.c [hunk 2]
```diff
@@ -519,6 +513,21 @@ ShadowStatus_t Shadow_MatchTopic( const char * pTopic,
                                                  pMessageType );
     }
 
+    if( shadowStatus == SHADOW_SUCCESS )
+    {
+        /* Update the out parameter if we successfully extracted the thing name. */
```
Comment (aggarg) [inline source/shadow.c:518]:
```suggestion
        /* Update the out parameters if we successfully matched the topic. */
```

File: source/shadow.c [hunk 3]
```diff
@@ -519,6 +513,21 @@ ShadowStatus_t Shadow_MatchTopic( const char * pTopic,
...
+        if( pThingName != NULL )
+        {
+            /* If match success, assign Thing Name which comes after shadow prefix. */
```
Comment (aggarg) [inline source/shadow.c:521]:
```suggestion
            /* Thing name comes after shadow prefix. */
```

File: test/unit-test/shadow_utest.c [hunk 4]
```diff
@@ -505,6 +505,24 @@ void test_Shadow_MatchTopic_Happy_Path( void )
...
+    TEST_ASSERT_EQUAL_INT( ShadowMessageTypeUpdateAccepted, messageType );
+
+    shadowStatus = Shadow_MatchTopic( &( topicBuffer[ 0 ] ),
+                                      bufferSize,
+                                      &messageType,
+                                      &pThingName,
+                                      NULL );
+    TEST_ASSERT_EQUAL_INT( SHADOW_SUCCESS, shadowStatus );
+    TEST_ASSERT_EQUAL_INT( ShadowMessageTypeUpdateAccepted, messageType );
+    TEST_ASSERT_EQUAL_STRING_LEN( TEST_THING_NAME, pThingName, TEST_THING_NAME_LENGTH );
```
[the remark this answers is not in the document]
Comment (DanielYEHsieh) [reply test/unit-test/shadow_utest.c:525]: added test case for both thing name and thing name length are NULL

commit 973e3f41f36f973aa708341b883ff52d8b1e8ca8

**Review by aggarg — APPROVED** [review]

**Review by sarenameas — APPROVED** [review]

**Pull request merged**